designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -72,77 +72,134 @@ class DeploymentServiceImpl(
-          case Failure(ex) =>
-            dbioRunner.runInTransaction(actionRepository.removeAction(actionId)).transform(_ => Failure(ex))
-          case Success(validationResult) =>
-            // we notify of deployment finish/fail only if initial validation succeeded
-            val deploymentFuture = runDeploymentActionWithNotifications(
-              actionType,
-              actionId,
-              processIdWithName,
-              versionOnWhichActionIsDone,
-              validatedComment,
-              buildInfoProcessIngType
-            ) {
-              dispatcher
-                .deploymentManagerUnsafe(processDetails.processingType)
-                .processCommand(
-                  RunDeploymentCommand(
-                    validationResult.processVersion,
-                    validationResult.deploymentData,
-                    validationResult.resolvedScenario,
-                    savepointPath
-                  )
-                )
-            }
-            Future.successful(deploymentFuture)
-        }
+        p => Some(p.processVersionId),
+        p => Some(p.processingType)
+      )
+      // 2. command specific section
+      deployedScenarioData <- prepareDeployedScenarioData(ctx.processDetails, ctx.actionId)
+      command = RunDeploymentCommand(
+        deployedScenarioData.processVersion,
+        deployedScenarioData.deploymentData,
+        deployedScenarioData.resolvedScenario,
+        savepointPath
+      )
+      // TODO: move validateBeforeDeploy before creating an action
+      result <- validateBeforeDeploy(ctx.processDetails, deployedScenarioData).transformWith {
+        case Failure(ex) =>
+          dbioRunner.runInTransaction(actionRepository.removeAction(ctx.actionId)).transform(_ => Failure(ex))
+        case Success(_) =>
+          // we notify of deployment finish/fail only if initial validation succeeded
+          val deploymentFuture = runActionAndHandleResults(
+            actionType,
+            ctx.actionId,
+            ctx.versionOnWhichActionIsDone,
+            validatedComment,
+            ctx.processDetails,
+            ctx.buildInfoProcessingType
+          ) {
+            dispatcher
+              .deploymentManagerUnsafe(ctx.processDetails.processingType)
+              .processCommand(command)
+          }
+          Future.successful(deploymentFuture)
       }
-    }
+    } yield result
+  }
+
+  /**
+    * Common validations and operations for a command execution.
+    * @return gathered data for further command execution
+    */
+  private def prepareCommandContextWithAction[PS: ScenarioShapeFetchStrategy](
+      processId: ProcessIdWithName,
+      actionType: ProcessActionType,
+      getVersionOnWhichActionIsDone: ScenarioWithDetailsEntity[PS] => Option[VersionId],
+      getBuildInfoProcessingType: ScenarioWithDetailsEntity[PS] => Option[ProcessingType]
+  )(implicit user: LoggedUser, ec: ExecutionContext): Future[CommandContext[PS]] = {
```

Comment:
are you sure that the for comprehension in deploy and cancel will work properly if you use Future here? I'm concerned about the situation that we had before, where failing the future would still run side effects. I think some testing of this method would be a good idea?

Reply:
Right now the only side effect is registered action and we have tests for that in DeploymentServiceSpec. What test scenario do you think of?

Reply:
I think if the registered action is covered in other tests then it's ok.